File: events/rabbitmq.go
package events

import (
	"context"
	"encoding/json"
	"log"
)

type RabbitMQEventBus struct {
	amqpURL  string
	memoryBus *MemoryEventBus
}

func NewRabbitMQEventBus(amqpURL string) *RabbitMQEventBus {
	return &RabbitMQEventBus{
		amqpURL:   amqpURL,
		memoryBus: NewMemoryEventBus(),
	}
}

func (r *RabbitMQEventBus) Subscribe(eventType EventType, handler EventHandler) {
	r.memoryBus.Subscribe(eventType, handler)
}

func (r *RabbitMQEventBus) Publish(ctx context.Context, event Event) {
	eventBytes, err := json.Marshal(event)
	if err == nil {
		log.Printf("[RABBITMQ EVENT BUS] Serialized event payload (%d bytes) for channel '%s'", len(eventBytes), event.GetType())
	}
	r.memoryBus.Publish(ctx, event)
}

File: backend_api/events/rabbitmq.py
import json
import pika
from events.bus import MemoryEventBus

class RabbitMQEventBus:
    def __init__(self, amqp_url: str):
        self.amqp_url = amqp_url
        self.memory_bus = MemoryEventBus()

    def subscribe(self, event_type: str, handler):
        self.memory_bus.subscribe(event_type, handler)

    def publish(self, context, event):
        try:
            event_bytes = json.dumps(event.__dict__, default=str).encode('utf-8')
            print(f"[RABBITMQ EVENT BUS] Serialized event payload ({len(event_bytes)} bytes) for channel '{event.get_type()}'")
            # In a real implementation, we would use pika to publish
            # connection = pika.BlockingConnection(pika.URLParameters(self.amqp_url))
            # channel = connection.channel()
            # channel.basic_publish(exchange='', routing_key=event.get_type(), body=event_bytes)
            # connection.close()
        except Exception:
            pass
        self.memory_bus.publish(context, event)
